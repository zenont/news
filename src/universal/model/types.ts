export type Language =
	'ar' | 'de' | 'en' | 'es' | 'fr' | 'he' | 'it' | 'nl' | 'no' | 'pt' | 'ru' | 'se' | 'ud' | 'zh'

export type SortBy =
	'relevancy' | 'popularity' | 'publishedAt'

export type Category =
	'business' | 'entertainment' | 'general' | 'health' | 'science' | 'sports' | 'technology'

export type Country =
	'ae' | 'ar' | 'at' | 'au' | 'be' | 'bg' | 'br' | 'ca' | 'ch' | 'cn' | 'de' | 'eg' | 'fr' | 'gb' | 'ru' | 'us'

export type Keywords = {
	exact?: ReadonlyArray<string>
	must?: ReadonlyArray<string>
	not?: ReadonlyArray<string>
}

export type UnhandledError = string | Error | null
