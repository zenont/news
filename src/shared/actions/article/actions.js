import { fetchSourcesAsync, fetchArticlesAsync } from '../../api'
import * as types from '../../constants'

export function fetchSourceOptionsAsync() {
	return (dispatch) => {
		dispatch({ type: types.NEWS_SOURCE_OPTIONS_REQUEST })

		return fetchSourcesAsync()
			.then(response => (response.json()), error => dispatch({ type: types.APP_SET_ERROR, payload: error }))
			.then(json => {
				const { sources } = json
				dispatch({
					type: types.NEWS_SOURCE_OPTIONS_RECEIVE,
					payload: {
						append: false,
						sources
					}
				})
			})
	}
}

export function fetchArticlesBySourceIdAsync(sourceId = 'cnn') {
	return (dispatch, getState) => {
		const { sources: { options: sourceOptions } } = getState().news.toJS()
		const selectedSource = sourceOptions.filter(sourceOption => sourceOption.id === sourceId)
		const { id = 'cnn', sortBysAvailables = ['top'] } = selectedSource
		console.log('fetchArticlesBySourceIdAsync', sourceId, id, selectedSource, sourceOptions)
		dispatch({ type: types.NEWS_ARTICLE_REQUEST })
		return fetchArticlesAsync(id, sortBysAvailables[0])
			.then(response => (response.json()), error => dispatch({ type: types.APP_SET_ERROR, payload: error }))
			.then(json => {
				const { articles } = json
				dispatch({
					type: types.NEWS_ARTICLE_RECEIVE,
					payload: {
						append: false,
						articles
					}
				})
			})
	}
}

export function selectSourceId(sourceId = 'cnn') {
	return { type: types.NEWS_SOURCE_OPTIONS_SELECT, payload: sourceId }
}
