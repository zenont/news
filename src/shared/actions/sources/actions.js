import { fetchSourcesAsync } from '../../api'

export function fetchNewsSourcesAsync() {
	return ({ dispatch }) => {
		return fetchSourcesAsync()
			.then(response => {
				if (response.ok) return response.json()
			}, error => console.warn(error))
			.then(json => {
				console.log('downloaded news sources', json)
			})
	}
}
