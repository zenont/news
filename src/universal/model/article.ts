
export class ArticleSource {
	public id?: string
	public name: string
}

export class Article {
	public source?: ArticleSource
	public author?: string
	public title: string
	public description: string
	public url: string
	public urlToImage?: string
	public publishedAt: Date
}
