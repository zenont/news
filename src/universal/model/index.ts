export * from './source'
export * from './article'
