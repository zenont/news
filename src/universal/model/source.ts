export type Source = {
	readonly id: string
	readonly name: string
	readonly description: string
	readonly url: string
	readonly category: string
	readonly language: string
	readonly country: string
}
