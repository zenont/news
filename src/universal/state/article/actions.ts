import { Action } from 'redux'
import { Article } from '../../model'

export enum ArticleActions {
	requestTopHeadlines = 'ARTICLES/REQUEST_TOP_HEADLINES',
	requestEverything = 'ARTICLES/REQUEST_EVERYTHING',
	fulfill = 'ARTICLES/FULFILL',
	reject = 'ARTICLES/REJECT',
	cancel = 'ARTICLES/CANCEL',
}

export type Language =
	'ar' | 'de' | 'en' | 'es' | 'fr' | 'he' | 'it' | 'nl' | 'no' | 'pt' | 'ru' | 'se' | 'ud' | 'zh'

export type SortBy =
	'relevancy' | 'popularity' | 'publishedAt'

export type Category =
	'business' | 'entertainment' | 'general' | 'health' | 'science' | 'sports' | 'technology'

export type Country =
	'ae' | 'ar' | 'at' | 'au' | 'be' | 'bg' | 'br' | 'ca' | 'ch' | 'cn' | 'de' | 'eg' | 'fr' | 'gb' | 'ru' | 'us'

export type Keywords = {
	exact?: ReadonlyArray<string>
	must?: ReadonlyArray<string>
	not?: ReadonlyArray<string>
}

export class TopHeadlinesRequest {
	public readonly country?: Country
	public readonly category?: Category
	public readonly sources?: ReadonlyArray<string>
	public readonly pageSize?: number
	public readonly page?: number
}

export class EverythingRequest {
	public readonly keywords?: Keywords
	public readonly sources?: ReadonlyArray<string>
	public readonly domains?: ReadonlyArray<string>
	public readonly from?: Date
	public readonly to?: Date
	public readonly language?: Language
	public readonly sortBy?: SortBy
	public readonly pageSize?: number
	public readonly page?: number
}

export class ArticleTopHeadlinesRequestAction extends TopHeadlinesRequest implements Action {
	public readonly type = ArticleActions.requestTopHeadlines
}

export class ArticleEverythingRequestAction implements Action {
	public readonly type = ArticleActions.requestEverything
}

export class IArticleFulfillAction implements Action {
	public readonly type = ArticleActions.fulfill
	constructor(
		public readonly articles: ReadonlyArray<Article>,
		public readonly total: number) { }
}

export class IArticleRejectAction implements Action {
	public readonly type = ArticleActions.reject
	constructor(public readonly error?: string | Error | null) { }
}

export class IArticleCancelAction implements Action {
	public readonly type = ArticleActions.cancel
}

export type ArticleAction =
	ArticleTopHeadlinesRequestAction |
	IArticleFulfillAction |
	IArticleRejectAction |
	IArticleCancelAction
