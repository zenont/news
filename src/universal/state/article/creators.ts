import { ArticleActions } from './types'
import { Article } from '../../model'
import * as actions from './actions'

export const requestArticles = (): actions.IArticleRequestAction => ({
	type: ArticleActions.request
})

export const requestTopHeadlines = (): actions.IArticleRequestTopHeadlinesAction => ({
	type: ArticleActions.requestTopHeadlines
})

export const fulfillArticles = (articles: Article[], total: number): actions.IArticleFulfillAction => ({
	type: ArticleActions.fulfill,
	articles,
	total
})

export const rejectArticles = (error: string | Error | any): actions.IArticleRejectAction => ({
	type: ArticleActions.reject,
	error
})

export const cancelArticles = (): actions.IArticleCancelAction => ({
	type: ArticleActions.cancel
})
