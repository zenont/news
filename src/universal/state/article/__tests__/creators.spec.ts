import * as A from '../actions'
import * as C from '../creators'
import { Article } from '../../../model'

describe('article creators', () => {
	it('should request', () => {
		const action = C.requestArticles()
		expect(action).toMatchSnapshot()
	})

	it('should fulfill', () => {
		const artiles: Article[] = [
			{
				author: 'author1',
				description: 'desc1',
				publishedAt: new Date(2000, 1),
				source: {
					id: '1',
					name: 'source1'
				},
				title: 'title1',
				url: 'url1',
				urlToImage: 'urlImage1'
			},
			{
				author: 'author2',
				description: 'desc2',
				publishedAt: new Date(2000, 2),
				source: {
					id: '2',
					name: 'source2'
				},
				title: 'title2',
				url: 'url2',
				urlToImage: 'urlImage2'
			}
		]

		const action = C.fulfillArticles(artiles, artiles.length)
		expect(action).toMatchSnapshot()
	})

	it('should reject with error message', () => {
		const action = C.rejectArticles('some error')
		expect(action).toMatchSnapshot()
	})

	it('should reject with Error', () => {
		const action = C.rejectArticles(new Error('some error'))
		expect(action).toMatchSnapshot()
	})

	it('should reject with any error', () => {
		const action = C.rejectArticles(1000)
		expect(action).toMatchSnapshot()
	})

	it('should cancel', () => {
		const action = C.cancelArticles()
		expect(action).toMatchSnapshot()
	})

	it('should request top headlines', () => {
		const action = C.requestTopHeadlines()
		expect(action).toMatchSnapshot()
	})
})
