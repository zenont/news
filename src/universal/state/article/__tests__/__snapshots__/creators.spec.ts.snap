// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`article creators should cancel 1`] = `
Object {
  "type": "ARTICLES/CANCEL",
}
`;

exports[`article creators should fulfill 1`] = `
Object {
  "articles": Array [
    Object {
      "author": "author1",
      "description": "desc1",
      "publishedAt": 2000-02-01T08:00:00.000Z,
      "source": Object {
        "id": "1",
        "name": "source1",
      },
      "title": "title1",
      "url": "url1",
      "urlToImage": "urlImage1",
    },
    Object {
      "author": "author2",
      "description": "desc2",
      "publishedAt": 2000-03-01T08:00:00.000Z,
      "source": Object {
        "id": "2",
        "name": "source2",
      },
      "title": "title2",
      "url": "url2",
      "urlToImage": "urlImage2",
    },
  ],
  "total": 2,
  "type": "ARTICLES/FULFILL",
}
`;

exports[`article creators should reject with Error 1`] = `
Object {
  "error": [Error: some error],
  "type": "ARTICLES/REJECT",
}
`;

exports[`article creators should reject with any error 1`] = `
Object {
  "error": 1000,
  "type": "ARTICLES/REJECT",
}
`;

exports[`article creators should reject with error message 1`] = `
Object {
  "error": "some error",
  "type": "ARTICLES/REJECT",
}
`;

exports[`article creators should request 1`] = `
Object {
  "type": "ARTICLES/REQUEST",
}
`;

exports[`article creators should request top headlines 1`] = `
Object {
  "type": "ARTICLES/REQUEST_TOP_HEADLINES",
}
`;
